temporalcli/commands.env.go

```diff
@@ -0,0 +1,87 @@
+package temporalcli
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/temporalio/cli/temporalcli/internal/printer"
+)
+
+func (c *TemporalEnvDeleteCommand) run(cctx *CommandContext, args []string) error {
+	keyPieces := strings.Split(args[0], ".")
+	if len(keyPieces) > 2 {
+		return fmt.Errorf("env property key to get cannot have more than one dot")
+	}
+	// Env must be present (but flag itself doesn't have to be)
+	env, ok := cctx.EnvConfigValues[keyPieces[0]]
+	if !ok {
+		return fmt.Errorf("env %q not found", keyPieces[0])
+	}
+	// User can remove single flag or all in env
+	if len(keyPieces) > 1 {
+		cctx.Logger.Info("Deleting env property", "env", keyPieces[0], "property", keyPieces[1])
+		delete(env, keyPieces[1])
+	} else {
+		cctx.Logger.Info("Deleting env", "env", keyPieces[0])
+		delete(cctx.EnvConfigValues, keyPieces[0])
+	}
+	return cctx.WriteEnvConfigToFile()
+}
+
+func (c *TemporalEnvGetCommand) run(cctx *CommandContext, args []string) error {
+	keyPieces := strings.Split(args[0], ".")
+	if len(keyPieces) > 2 {
+		return fmt.Errorf("env key to get cannot have more than one dot")
+	}
+	// Env must be present (but flag itself doesn't have to me)
+	env, ok := cctx.EnvConfigValues[keyPieces[0]]
+	if !ok {
+		return fmt.Errorf("env %q not found", keyPieces[0])
+	}
+	type prop struct {
+		Property string `json:"property"`
+		Value    string `json:"value"`
+	}
+	var props []prop
+	// User can ask for single flag or all in env
+	if len(keyPieces) > 1 {
+		props = []prop{{Property: keyPieces[1], Value: env[keyPieces[1]]}}
+	} else {
+		props = make([]prop, 0, len(env))
+		for k, v := range env {
+			props = append(props, prop{Property: k, Value: v})
+		}
+		sort.Slice(props, func(i, j int) bool { return props[i].Property < props[j].Property })
+	}
+	// Print as table
+	return cctx.Printer.PrintStructured(props, printer.StructuredOptions{Table: &printer.TableOptions{}})
+}
+
+func (c *TemporalEnvListCommand) run(cctx *CommandContext, args []string) error {
+	type env struct {
+		Name string `json:"name"`
+	}
+	envs := make([]env, 0, len(cctx.EnvConfigValues))
+	for k := range cctx.EnvConfigValues {
+		envs = append(envs, env{Name: k})
+	}
+	// Print as table
+	return cctx.Printer.PrintStructured(envs, printer.StructuredOptions{Table: &printer.TableOptions{}})
```

Comment:
> You had concern for script use before, but this would be unsafe/incompatible for a dev who wants to add a second column. There is value in consistency here I think, but I can change if the look of a single column header is more offputting than inconsistency.

Changing the number of columns in the output is itself a breaking change, so I think changing whether headers are output between 1-column and N-column output is a moot point. Your other points/questions about table output are valid, however, and I think can all be addressed by adding a global option to allow users to omit table headers.

> Note: This isn't a heavily used command yet I see we're applying deep scrutiny to its text output and JSON names and such, but the really heavily used commands don't seem to be getting similar scrutiny.

I don't agree; I ran all the commands and inspected their output. So perhaps it is more that I applied equal scrutiny to every command rather than more scrutiny to important commands. I will double-check the `workflow` command output though if it makes you feel better.

Reply:
I went back over the `workflow` commands and I don't have strong opinions about their output, other than it would be nice to have an option to print only the result of the workflow (on commands like `execute` and `describe`). That's an enhancement, however, and is out of scope for this PR.

So I think all that's left to resolve this comment is: Do we want to add a global option to omit the table headers? (Perhaps this could be a post-rewrite, pre-1.0 ticket if we want to do it. Though to strip the headers, `temporal ... | tail +2` is not hard for users who know that's a thing they can do.)

Reply:
> Changing the number of columns in the output is itself a breaking change

Not necessarily. Adding headers sometimes and not others is much more jarring and inconsistent than adding columns to an existing table with headers.

> I think changing whether headers are output between 1-column and N-column output is a moot point.

It's not. Surely it makes sense that adding columns to a table is less jarring than showing/hiding headers. If I know table data always has a header, I can prepare for that. But if there's some weird Temporal rules where some tables have headers and some don't, that's pretty inconsistent.

> I think can all be addressed by adding a global option to allow users to omit table headers.

True, but I think it's overthinking. Should an option be present to change from center to left align? Magenta to red? I think we should avoid considering text output as script consumable or we're going to have a lot of problems. Note that the columns still have alignment and indentation which isn't as script compatible.

> Do we want to add a global option to omit the table headers? 

I don't think so, because it only makes sense on single-table output, and by that logic we can have a ton of configuration options about tables, headers is but one. At the least we should wait for a real, demonstrable need before adding these things now.

Reply:
> Surely it makes sense that adding columns to a table is less jarring than showing/hiding headers.

Completely disagree—adding/changing/removing columns is _far_ more jarring than showing/hiding headers in basically every situation. It is almost _always_ a breaking change for scripts using the text output (with the possible exception of adding new columns on the end, when you already have a multi-column table).

> > Do we want to add a global option to omit the table headers?
>
> I don't think so, because it only makes sense on single-table output with single column (many commands have multiple tables and/or multiple columns), and by that logic we can have a ton of configuration options about tables, headers is but one.

I think it also makes sense even on multi-column tables, when script writers might want to parse the table but already know what the headers are. I do not think there is value in e.g. alignment options because most people will just account for arbitrary spacing between columns (e.g. `(\S+)\s+(\S+)...`) and call it good.

> At the least we should wait for a real, demonstrable need before adding these things now.

Agree it does not need to be done in this PR. I don't think we should wait too long though.

Reply:
Filed https://github.com/temporalio/cli/issues/419 for this.